discovery-provider/src/tasks/index_aggregate_plays.py

```diff
@@ -1,17 +1,101 @@
 import logging
 import time
+import sqlalchemy as sa
 from src.tasks.celery_app import celery
+from src.models import IndexingCheckpoints, Play
 
 logger = logging.getLogger(__name__)
 
+AGGREGATE_PLAYS_TABLE_NAME = "aggregate_plays"
 
-def update(self, db):
-    with db.scoped_session() as session:
-        start_time = time.time()
-        session.execute("REFRESH MATERIALIZED VIEW CONCURRENTLY aggregate_plays")
+### UPDATE_AGGREGATE_PLAYS_QUERY ###
+# Get new plays that came after the last indexing checkpoint for aggregate_play
+# Group those new plays by play item id to get the aggregate play counts
+# For new play item ids, insert those aggregate counts
+# For existing play item ids, add the new aggregate count to the existing aggregate count
+UPDATE_AGGREGATE_PLAYS_QUERY = """
+    WITH aggregate_plays_last_index AS (
+        SELECT
+            :prev_id_checkpoint AS prev_id_checkpoint,
+            :new_id_checkpoint AS new_id_checkpoint
+    ),
+    new_plays AS (
+        SELECT
+            play_item_id,
+            count(play_item_id) AS count
+        FROM
+            plays p
+        WHERE
+            p.id > (
+                SELECT
+                    prev_id_checkpoint
+                FROM
+                    aggregate_plays_last_index
+            )
+            AND p.id <= (
+                SELECT
+                    new_id_checkpoint
+                FROM
+                    aggregate_plays_last_index
+            )
+        GROUP BY
+            play_item_id
+    )
+    INSERT INTO
+        aggregate_plays (play_item_id, count)
+    SELECT
+        new_plays.play_item_id,
+        new_plays.count
+    FROM
+        new_plays ON CONFLICT (play_item_id) DO
+    UPDATE
+    SET
+        count = aggregate_plays.count + EXCLUDED.count
+    """
+
+UPDATE_INDEXING_CHECKPOINTS = """
+    INSERT INTO indexing_checkpoints (tablename, last_index)
+    VALUES(:tablename, :last_index)
+    ON CONFLICT (tablename)
+    DO UPDATE SET last_index = EXCLUDED.last_index;
+    """
+
+def _update_aggregate_plays(session):
+    # get the last updated id that counted towards the current aggregate plays
+    prev_id_checkpoint = (session.query(IndexingCheckpoints.last_index)
+        .filter(IndexingCheckpoints.tablename == AGGREGATE_PLAYS_TABLE_NAME)
+    ).scalar()
+
+    if not prev_id_checkpoint:
+        prev_id_checkpoint = 0
+
+    # get the new latest
+    new_id_checkpoint = (
+        session.query(Play.id)
+        .order_by(Play.id.desc()).limit(1)
+    ).scalar()
+
+    if not new_id_checkpoint:
+        new_id_checkpoint = 0
```

Comment:
should we explicitly handle the two cases where new is 0?
- if prev != 0 and new = 0 this is corrupted state
- if prev = 0 and new = 0, it's a no-op we can probably early exit

Reply:
yeah i considered using one() instead of scalar() so it throws an exception. it should never happen but i guess it's good to have. i'll add it.